### PR TITLE
Fix Manifold<3, 3>::normal_vector

### PR DESCRIPTION
**Pull request by zjiaqi2018**
fix  https://github.com/dealii/dealii/issues/14176
On very coarse meshes, with very particular manifolds, e.g., spherical manifold, polar manifolds, cylindrical manifolds, etc, `Manifold<3, 3>::normal_vector` may return a zero vector. This issues can be fixed by choosing different combinations of vertices to compute the normal vector.

### CODE REVIEW

File: include/deal.II/grid/grid_generator.h [hunk 1]
```diff
@@ -1427,9 +1427,10 @@ namespace GridGenerator
    * <code>false</code> all indicators are set to 0 (see
    * @ref GlossColorization "the glossary entry on colorization").
    *
-   * All manifold ids are set to zero, and a SphericalManifold is attached
-   * to the triangulation.
-   *
+   * In 2d, all manifold ids are set to zero, and a SphericalManifold is
+   * attached to the triangulation. In 3d, only the manifold id of the curved
```
Comment (tjhei) [inline include/deal.II/grid/grid_generator.h:1431]: Why is this different in 2d? I would think there are also flat faces in 2d!

Comment (zjiaqi2018) [reply include/deal.II/grid/grid_generator.h:1431]: Then we may also need to fix quarter_hyper_shell()?https://github.com/dealii/dealii/blob/d7aca55de557876b24d95d89fc02f82da4d8d271/source/grid/grid_generator.cc#L6313

Comment (masterleinad) [reply include/deal.II/grid/grid_generator.h:1431]: I would think so. Can you try to address that as well and see what happens?

Comment (zjiaqi2018) [reply include/deal.II/grid/grid_generator.h:1431]: If we remove the spherical manifold on the flat surface, the mesh quality there gets worse as we refine the mesh, as discussed here https://github.com/dealii/dealii/issues/14176

**Review by masterleinad — CHANGES_REQUESTED** [review]
This is not the right solution, see https://github.com/dealii/dealii/issues/14176#issuecomment-1212477309.

**Comment by zjiaqi2018** [comment]
It should work now, @tjhei @luca-heltai

File: source/grid/manifold.cc [hunk 2]
```diff
@@ -206,9 +206,45 @@ Manifold<3, 3>::normal_vector(const Triangulation<3, 3>::face_iterator &face,
...
+  tested_vertices[second_index] = true;
+
+  do
+    {
+      // Compute tangents and normal for selected vertices
+      t1     = get_tangent_vector(p, vertices[first_index]);
+      t2     = get_tangent_vector(p, vertices[second_index]);
+      normal = cross_product_3d(t1, t2);
+
+      // if normal is zero, try some other combination of veritices
```
Comment (tjhei) [inline source/grid/manifold.cc:225]: Typo vertices

File: source/grid/manifold.cc [hunk 3]
```diff
@@ -206,9 +206,45 @@ Manifold<3, 3>::normal_vector(const Triangulation<3, 3>::face_iterator &face,
...
+      if (normal.norm_square() < 1e4 * std::numeric_limits<double>::epsilon() *
+                                   t1.norm_square() * t2.norm_square())
+        {
+          // See if we have tested already all vertices
```
Comment (tjhei) [inline source/grid/manifold.cc:229]:
```suggestion
          // See if we have tested all vertices already 
```

**Review by tjhei — APPROVED** [review]

**Comment by tjhei** [comment]
Can you add a changelog and update the text of the PR (when you scroll up here)?

**Review by luca-heltai — APPROVED** [review]

**Review by masterleinad — APPROVED** [review]

**Pull request merged**